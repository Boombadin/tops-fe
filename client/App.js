import React from 'react';
import { Switch, Route } from 'react-router-dom';
import routes from './routes';
import uuidV4 from 'uuid/v4';

const App = () => (
  <Switch>
    {routes.map(route => (
      <Route key={uuidV4()} {...route} />
    ))}
  </Switch>
);

export default App;
