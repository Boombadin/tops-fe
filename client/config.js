export const googleCaptcha = {
  sitekey: '6Lc5OlUUAAAAAL2IPe0wDYCo3OOxjqYHNAFU5Eo6',
};

export const sentryClient = {
  dsn: 'https://54fca34e786c4f47b259c2d041a84dc3@o118326.ingest.sentry.io/5323535',
  env: 'staging',
  version: '3.1.0',
};
