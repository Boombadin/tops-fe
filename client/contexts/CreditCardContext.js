import {
  useCreateCardMutation,
  useDeleteCardMutation,
  useSetDefaultCardMutation,
} from '@central-tech/react-hooks';
import React, { useContext, useEffect, useRef, useState } from 'react';

import { usePaymentContext } from '@client/contexts';
import CreditCardModel from '@client/model/CreditCard/CreditCardModel';
import { useCardsQuery } from '@client/operations';

import { creditCardEventType } from '../constants/creditCardEventType';

const CreditCardContext = React.createContext();

const CreditCardProvider = ({ children }) => {
  const refCreditCardFrame = useRef(null);
  const [isLoading, setIsLoading] = useState(true);
  const [isShowDeleteCardModal, setShowDeleteCardModal] = useState(false);
  const [isSetDefaultCreditCard, setIsSetDefaultCreditCard] = useState(true);
  const [cardList, setCardList] = useState(null);
  const [currentSelectedCard, setCurrentSelectedCard] = useState(null);
  const [newCard, setNewCard] = useState(null);
  const [isInvalidCard, setIsInvalidCard] = useState(null);
  const [
    currentSelectedCardToDelete,
    setCurrentSelectedCardToDelete,
  ] = useState();
  const [isShowCardList, setIsShowCardList] = useState(true);
  const [
    openCreditCardFormInSavedList,
    setOpenCreditCardFormInSavedList,
  ] = useState(false);
  const [isCreditCardFormActive, setCreditCardFormActive] = useState(false);
  const [configDiscardFormModal, setConfigDiscardFormModal] = useState({
    isOpen: false,
    navigatePath: null,
  });
  const [
    currentSelectedCardPromotion,
    setCurrentSelectedCardPromotion,
  ] = useState('');
  const [isShowCardPromotion, setIsShowCardPromotion] = useState(false);
  const [isCardPromotionLoading, setIsCardPromotionLoading] = useState(false);

  const { paymentAction } = usePaymentContext();

  // Core API
  const refetchQueries = [`cards`];
  const [createCardMutation, createCardResponse] = useCreateCardMutation({
    refetchQueries,
  });
  const [setDefaultCardMutation] = useSetDefaultCardMutation();
  const [deleteCardMutation] = useDeleteCardMutation({
    refetchQueries,
  });
  const cardsResponse = useCardsQuery({
    fetchPolicy: 'no-cache',
    variables: {
      sort: {
        id: 'CREATED_AT',
        direction: 'ASC',
      },
    },
  });
  const { data, loading } = cardsResponse;

  useEffect(() => {
    if (isShowCardList) {
      setCurrentSelectedCard(getDefaultCard());
    } else {
      setCurrentSelectedCard(null);
    }
  }, [isShowCardList]);

  useEffect(() => {
    if (currentSelectedCard) {
      paymentAction.setPaymentServiceFullPaymentSavedCard({
        savedCard: { ...currentSelectedCard },
      });
    } else {
      paymentAction.setPaymentServiceFullPaymentSavedCard({
        savedCard: null,
      });
    }
  }, [currentSelectedCard]);

  useEffect(() => {
    setIsLoading(loading);
    const cards = data?.cards || [];
    if (cards?.length > 0) {
      const formattedCardList = cards.map(
        card => new CreditCardModel({ card }),
      );
      const activeCardList = formattedCardList?.filter(
        card => !card?.isExpired,
      );
      setCardList(activeCardList);
    } else {
      setCardList([]);
      setCurrentSelectedCard(null);
    }
  }, [data, loading, newCard]);

  useEffect(() => {
    // first time call
    if (cardList?.length > 0) {
      const tempDefaultCard = getDefaultCard();
      if (newCard) {
        setCurrentSelectedCard(getLastSavedCard(newCard.id));
        setNewCard(null);
      } else {
        setCurrentSelectedCard(tempDefaultCard);
      }
    }
  }, [cardList]);

  const submitNewCard = async card => {
    try {
      await createCardMutation({
        variables: {
          cardInput: {
            encrypted_card_data: card?.encryptedCardInfo,
            cardholder_name: card?.cardHolderName,
            expiry_month: parseInt(card?.expMonthCardInfo),
            expiry_year: parseInt(card?.expYearCardInfo),
            is_store_card: true,
          },
          setDefault: isSetDefaultCreditCard || cardList?.length === 0,
        },
      });
    } catch (e) {
      setIsLoading(false);
      console.error(e);
    }
  };

  useEffect(() => {
    const card = createCardResponse?.data?.createCard;
    if (card) {
      setNewCard(
        new CreditCardModel({
          card,
        }),
      );
      setIsShowCardList(true);
      setIsLoading(false);
    }
    if (createCardResponse?.error?.message) setIsInvalidCard(true);
  }, [createCardResponse]);

  useEffect(() => {
    // reset state
    if (isShowCardList && isInvalidCard) setIsInvalidCard(null);
    if (isShowCardList && isSetDefaultCreditCard)
      setIsSetDefaultCreditCard(true);
  }, [isShowCardList]);

  const handleSubmitNewCard = async card => {
    setIsLoading(true);
    await submitNewCard(card);
  };

  const handleSetDefaultCardIconClick = card => {
    setIsLoading(true);
    setDefaultCardMutation({
      variables: {
        cardId: card.id,
      },
    })
      .then(({ data }) => {
        if (data?.setDefaultCard) {
          cardList?.map(
            cardItem =>
              (cardItem.isDefault = cardItem.id === data?.setDefaultCard?.id),
          );
        }
      })
      .finally(() => setIsLoading(false));
  };

  const handleDeleteCardIconClick = card => {
    setCurrentSelectedCardToDelete(card);
    setShowDeleteCardModal(true);
  };

  const handleOnConfirmDeleteCard = () => {
    setShowDeleteCardModal(false);
    setIsLoading(true);
    deleteCardMutation({
      variables: {
        cardId: currentSelectedCardToDelete.id,
      },
    })
      .then()
      .finally(() => setIsLoading(false));
  };

  const getDefaultCard = () => cardList?.find(card => card?.isDefault) || null;
  const getLastSavedCard = id =>
    cardList?.find(card => String(card?.id) === String(id));
  const isCardsLoading = loading;
  const isEmptyCards = cardList !== null && cardList?.length === 0;

  const setOpenDiscardNewCreditCardFormModal = ({
    open = true,
    navigatePath = '',
  }) => {
    if (open) {
      setConfigDiscardFormModal(prevState => ({
        ...prevState,
        isOpen: true,
        navigatePath,
      }));
    } else {
      setConfigDiscardFormModal(prevState => ({
        ...prevState,
        isOpen: false,
        navigatePath: '',
      }));
    }
  };

  const value = {
    cardList,
    isLoading,
    isCardsLoading,
    isEmptyCards,
    refCreditCardFrame,
    currentSelectedCard,
    setCurrentSelectedCard,
    defaultCard: getDefaultCard(),
    isShowDeleteCardModal,
    setShowDeleteCardModal,
    handleSetDefaultCardIconClick,
    handleDeleteCardIconClick,
    handleOnConfirmDeleteCard,
    handleSubmitNewCard,
    isSetDefaultCreditCard,
    setIsSetDefaultCreditCard,
    isShowCardList,
    setIsShowCardList,
    currentSelectedCardToDelete,
    isInvalidCard,
    openCreditCardFormInSavedList,
    setOpenCreditCardFormInSavedList,
    isCreditCardFormActive,
    setCreditCardFormActive,
    setOpenDiscardNewCreditCardFormModal,
    configDiscardFormModal,
    setCurrentSelectedCardPromotion,
    currentSelectedCardPromotion,
    isShowCardPromotion,
    setIsShowCardPromotion,
    isCardPromotionLoading,
    setIsCardPromotionLoading,
  };

  return (
    <CreditCardContext.Provider value={value}>
      {children}
    </CreditCardContext.Provider>
  );
};

const useCreditCardContext = () => useContext(CreditCardContext);

export { CreditCardContext, CreditCardProvider, useCreditCardContext };
