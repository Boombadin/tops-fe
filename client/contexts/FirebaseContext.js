import * as firebase from 'firebase/app';
import isEmpty from 'lodash/isEmpty';
import React, {
  createContext,
  useCallback,
  useContext,
  useEffect,
  useState,
} from 'react';

import setFirebase from '@client/config/firebase';

import 'firebase/remote-config';
import 'firebase/performance';
import 'firebase/firestore';

// Init config Firebase
const env = window?.App?.environment;
let setupFirebase = setFirebase.staging;

if (env === 'prod') {
  setupFirebase = setFirebase.prod;
} else if (env === 'uat') {
  setupFirebase = setFirebase.uat;
}

const {
  apiKey,
  authDomain,
  databaseURL,
  projectId,
  storageBucket,
  messagingSenderId,
  appId,
} = setupFirebase.config;

const firebaseConfig = {
  apiKey: apiKey,
  authDomain: authDomain,
  databaseURL: databaseURL,
  projectId: projectId,
  storageBucket: storageBucket,
  messagingSenderId: messagingSenderId,
  appId: appId,
};

export const FirebaseContext = createContext();

export function useFirebaseContext() {
  return useContext(FirebaseContext);
}

function FirebaseProvider({ children }) {
  const [remoteConfig, setRemoteConfig] = useState(null);
  const [firestore, setFirestore] = useState(null);

  const getSeasonalConfig = useCallback(async () => {
    if (firestore) {
      const seasonalConfigDoc = await firestore
        .collection('seasonal')
        .doc('config')
        .get();

      if (seasonalConfigDoc.exists) {
        return seasonalConfigDoc.data();
      }
    }

    return {};
  }, [firestore]);

  const getFireStoreConfig = useCallback(
    async fireStoreConfig => {
      if (firestore) {
        const fireStoreConfigDoc = await firestore
          .collection(fireStoreConfig?.collection)
          .doc(fireStoreConfig?.doc)
          .get();

        if (fireStoreConfigDoc.exists) {
          return fireStoreConfigDoc.data();
        }
      }

      return {};
    },
    [firestore],
  );

  const getRemoteConfig = useCallback(
    async key => {
      const remoteValue = await remoteConfig.fetchAndActivate().then(() => {
        return remoteConfig.getValue(key).asString();
      });
      if (!isEmpty(remoteValue)) {
        return remoteValue;
      }
      return '';
    },
    [remoteConfig],
  );

  useEffect(() => {
    // Initialize Firebase
    firebase.initializeApp(firebaseConfig);
    // Initialize Performance Monitoring and get a reference to the service
    // firebase.performance();

    const firebaseRemoteConfig = firebase.remoteConfig();
    firebaseRemoteConfig.settings = {
      minimumFetchIntervalMillis: 100,
    };
    setRemoteConfig(firebaseRemoteConfig);
    setFirestore(firebase.firestore());
  }, []);

  const remoteConfigStore = {
    remoteConfig,
    firestore,
    firestoreAction: {
      getSeasonalConfig,
      getRemoteConfig,
      getFireStoreConfig,
    },
  };

  return (
    <FirebaseContext.Provider value={remoteConfigStore}>
      {children}
    </FirebaseContext.Provider>
  );
}

export default FirebaseProvider;
