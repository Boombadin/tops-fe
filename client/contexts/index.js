export {
  default as CartProvider,
  useCartContext,
  CartContext,
} from '@client/contexts/CartContext';
export {
  CreditCardProvider,
  useCreditCardContext,
  CreditCardContext,
} from '@client/contexts/CreditCardContext';
export {
  default as FirebaseProvider,
  useFirebaseContext,
  FirebaseContext,
} from '@client/contexts/FirebaseContext';
export {
  default as PaymentProvider,
  usePaymentContext,
  PaymentContext,
} from '@client/contexts/PaymentContext';
export {
  default as ReduxProvider,
  useReduxContext,
  ReduxContext,
} from '@client/contexts/ReduxContext';
