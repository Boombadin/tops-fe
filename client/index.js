import { CoreUiProvider } from '@central-tech/core-ui';
import * as Sentry from '@sentry/browser';
import axios from 'axios';
import Cookies from 'js-cookie';
import { find } from 'lodash';
import React from 'react';
import ReactDOM from 'react-dom';
import { addTranslation } from 'react-localize-redux';
import { Provider } from 'react-redux';
import { BrowserRouter as Router } from 'react-router-dom';

import { createBrowserApolloClient } from '@/createGraphqlClient';
import translation from '@/translate/translation.json';
import App from '@client/App';
import theme from '@client/config/theme';
import {
  CartProvider,
  FirebaseProvider,
  ReduxProvider,
} from '@client/contexts';
import CreditCardFrame from '@client/pages/CreditCardFrame';
import { fetchCart } from '@client/reducers/cart';
import { CreateSentryClient } from '@client/sentryClient';
import generateStore from '@client/store';

import './styles/app.scss';

const window_app = window?.App;
const store = generateStore(window_app?.state);
store.dispatch(addTranslation(translation));
const RootCreditCardFrame = document.getElementById('root-credit-card-frame');

class RootApp extends React.Component {
  componentDidCatch(error, errorInfo) {
    if (typeof window !== 'undefined') {
      window?.newrelic?.noticeError(error);
    }
    if (true) {
      Sentry.withScope(scope => {
        scope.setExtras(errorInfo);
        Sentry.captureException(error);
      });
    }
  }

  render() {
    if (RootCreditCardFrame) {
      return (
        <Provider store={store}>
          <CreditCardFrame />
        </Provider>
      );
    }

    const state = store.getState();

    const storeCode = state.storeConfig.current.code;
    const siteId = state.storeConfig.current.website_id;
    const isLogin = Object.keys(state.customer).length > 0;
    const currentLang = find(state.locale.languages, { active: true });

    axios.defaults.headers.common['x-store-code'] = storeCode;
    axios.defaults.headers.common['x-website-id'] = siteId;

    if (isLogin) {
      store.dispatch(fetchCart());
    }
    CreateSentryClient();

    const userToken = Cookies.get('user_token');
    const { client } = createBrowserApolloClient({
      uri: window_app?.graphql_url,
      headers: {
        store: storeCode,
        bu: 'tops',
        client: 'web',
        authorization: userToken ? `Bearer ${userToken}` : '',
      },
    });
    const context = {
      store,
      client,
      storeConfig: window_app?.state?.storeConfig?.current,
      lang: currentLang?.url,
      customer: window_app?.state?.customer,
      userToken: userToken || null,
    };

    return (
      <Provider store={context?.store}>
        <CoreUiProvider client={context?.client} theme={theme}>
          <Router basename={`/${currentLang.url}`}>
            <ReduxProvider store={store}>
              <FirebaseProvider>
                <CartProvider>
                  <App />
                </CartProvider>
              </FirebaseProvider>
            </ReduxProvider>
          </Router>
        </CoreUiProvider>
      </Provider>
    );
  }
}

ReactDOM.hydrate(
  <RootApp />,
  RootCreditCardFrame
    ? document.getElementById('root-credit-card-frame')
    : document.getElementById('root'),
);
