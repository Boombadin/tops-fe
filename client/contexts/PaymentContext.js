import { useUpdatePaymentInformationMutation } from '@central-tech/react-hooks';
import Cookie from 'js-cookie';
import PropTypes from 'prop-types';
import React, {
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useReducer,
  useRef,
  useState,
} from 'react';

import { useReduxContext } from '@client/contexts';
import paymentActions from '@client/contexts/payment/actions';
import PaymentModel from '@client/contexts/payment/model';
import reducer from '@client/contexts/payment/reducer';
import {
  typesPaymentMethods,
  typeUpdatePayment,
} from '@client/contexts/payment/types';
import {
  usePaymentMethodsQuery,
  useSetPaymentInformationMutation,
} from '@client/operations';

const mutationSetting = {
  awaitRefetchQueries: true,
};

export const PaymentContext = React.createContext();

export function usePaymentContext() {
  return useContext(PaymentContext);
}

const initialState = {
  paymentCode: '',
  paymentLoading: false,
  paymentServiceFullPayment: {
    promotionCode: '',
    saveCard: true,
    savedCard: null,
    isInvalidCard: false,
  },
};

function PaymentProvider({ children }) {
  const [payment, setPayment] = useState(new PaymentModel({ payment: {} }));

  const { reduxState, reduxAction } = useReduxContext();
  const cart = useMemo(() => reduxState.cart?.cart, [reduxState.cart?.cart]);
  const currentStoreCode = useMemo(
    () => reduxState.storeConfig?.current?.code,
    [reduxState.storeConfig],
  );

  const refCreditCardFrame = useRef(null);
  const refCvvForm = useRef(null);

  const others = useMemo(
    () => ({
      refCreditCardFrame,
      refCvvForm,
    }),
    [refCreditCardFrame, refCvvForm],
  );

  const [state, dispatch] = useReducer(reducer, initialState);
  const paymentMethodResponse = usePaymentMethodsQuery({
    variables: {
      isGuest: false,
    },
    headers: {
      Authorization: `Bearer ${Cookie.get('user_token')}`,
      store: currentStoreCode,
    },
    fetchPolicy: 'no-cache',
    notifyOnNetworkStatusChange: true,
  });

  /*** setPaymentInformation for create order ***/
  const [
    setPaymentInformationMutation,
    setPaymentInformationResponse,
  ] = useSetPaymentInformationMutation();
  const setPaymentInformation = useCallback(
    ({ variables }) => {
      return setPaymentInformationMutation({
        ...mutationSetting,
        variables: {
          isGuest: false,
          ...variables,
        },
        headers: {
          Authorization: `Bearer ${Cookie.get('user_token')}`,
          store: currentStoreCode,
        },
      });
    },
    [setPaymentInformationMutation, currentStoreCode],
  );

  /***  set payment code and promotion code ***/
  const [
    updatePaymentInformationMutation,
    updatePaymentInformationResponse,
  ] = useUpdatePaymentInformationMutation();

  const updatePaymentInformation = useCallback(
    async ({ type, promotionCode, paymentCode }) => {
      const refetchQueries = ['paymentMethods', 'cart'];

      const input = {
        payment_method: {
          method: paymentCode,
          extension_attributes: {},
        },
        email: cart?.customer?.email,
      };

      switch (type) {
        case typeUpdatePayment.SET_PAYMENT_METHOD:
          const listPaymentService = [
            typesPaymentMethods.PAYMENT_SERVICE_FULLPAYMENT,
          ];
          if (listPaymentService.includes(paymentCode)) {
            input.payment_service_methods = paymentCode;
          }
          break;
        case typeUpdatePayment.SET_PROMOTION_CODE:
          input.payment_method.extension_attributes = {
            promotion_id: String(promotionCode),
          };

          break;
        default:
          console.error('updatePaymentInformation: No type');
      }

      return updatePaymentInformationMutation({
        ...mutationSetting,
        variables: {
          isGuest: false,
          input,
        },
        headers: {
          Authorization: `Bearer ${Cookie.get('user_token')}`,
          store: currentStoreCode,
        },
        refetchQueries,
      });
    },
    [cart?.customer?.email, updatePaymentInformationMutation, currentStoreCode],
  );

  const paymentAction = paymentActions(
    { payment, cart, state },
    dispatch,
    { reduxState, reduxAction },
    {
      setPaymentInformation,
      updatePaymentInformation,
      others,
    },
  );

  useEffect(() => {
    if (paymentMethodResponse?.data?.paymentInformations) {
      setPayment(
        new PaymentModel({
          payment: paymentMethodResponse?.data?.paymentInformations,
        }),
      );
    }
  }, [paymentMethodResponse?.data?.paymentInformations]);

  useEffect(() => {
    if (
      paymentMethodResponse?.loading ||
      updatePaymentInformationResponse?.loading ||
      setPaymentInformationResponse?.loading
    ) {
      paymentAction.setPaymentLoading(true);
    } else {
      paymentAction.setPaymentLoading(false);
    }
  }, [
    paymentMethodResponse?.loading,
    updatePaymentInformationResponse?.loading,
    setPaymentInformationResponse?.loading,
  ]);

  const paymentStore = useMemo(
    () => ({
      payment: { ...state, ...payment, ...others, typesPaymentMethods },
      paymentAction,
    }),
    [state, payment, others, paymentAction],
  );

  return (
    <PaymentContext.Provider value={paymentStore}>
      {children}
    </PaymentContext.Provider>
  );
}

PaymentProvider.propTypes = {
  children: PropTypes.node.isRequired,
};

export default PaymentProvider;
