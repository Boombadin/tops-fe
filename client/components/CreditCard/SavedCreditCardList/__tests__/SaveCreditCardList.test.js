import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';

import { useCreditCardContext, useReduxContext } from '@/client/contexts';

import mockCardList from '../__mocks__/cardList';
import SaveCreditCardList from '../SavedCreditCardList';

jest.mock(
  '@client/components/CreditCard/SavedCreditCardListItem',
  () => props => (
    <mock-creditCardItem
      {...props}
      onClick={props.onSelectCard}
      data-testid="inf-viewSavedCreditCardListItem-Container"
    />
  ),
);
jest.mock('@client/components/CreditCard/TooltipCvv', () => () => (
  <mock-tooltipCvv />
));
jest.mock('@client/components/CreditCard/NewCreditCardForm', () => () => (
  <mock-newCreditCardForm />
));
jest.mock('@client/hoc/withLocales', () => {
  return Component => props => (
    <Component translate={jest.fn(key => key)} {...props} />
  );
});
jest.mock('@client/contexts', () => ({
  useFirebaseContext: jest.fn().mockReturnValue({
    firestoreAction: {
      getRemoteConfig: jest.fn().mockReturnValue(
        JSON.stringify({
          KCC: { th: 'ธนาคารกรุงศรีอยุธยา', en: 'Bank of Ayudhya' },
          BBL: { th: 'ธนาคารกรุงเทพ', en: 'Bangkok Bank' },
          Citibank: { th: 'ธนาคาร ซิตี้ แบงก์', en: 'Citibank' },
          FirstChoice: {
            th: 'กรุงศรีเฟิร์สช้อยส์',
            en: 'Krungsri First Choice',
          },
        }),
      ),
    },
  }),
  useReduxContext: jest.fn().mockReturnValue({}),
  usePaymentContext: jest.fn().mockReturnValue({}),
  useCreditCardContext: jest.fn().mockReturnValue({}),
}));

const TEST_ID = {
  CVV_FORM_CONTAINER: 'frm-viewSavedCreditCardList-CvvFormContainer',
  CVV_FORM: 'frm-viewSavedCreditCardList-CvvForm',
  INPUT_CVV_FORM: 'inp-formSavedCreditCardList-CardCVV',
  CARD_CVV_ERROR: 'inp-formSavedCreditCardList-CardCVVError',
  ADD_NEW_CARD: 'btn-addSavedCreditCardList-AddNewCard',
};

const TEST_DATA = {
  ERRORS_CVV: {
    CARDCVV_REQUIRED: 'CVV is required',
    CARDCVV_INVALID: 'Invalid CVV',
  },
};
describe('components/CreditCard/SavedCreditCardList', () => {
  const props = {
    handleSelectCard: jest.fn(),
  };
  const mockCreditCardContext = {
    cardList: mockCardList,
    currentSelectedCard: {
      bankIcon: '/images/payment/product-img-empty.svg',
      bankId: null,
      bankImage: '/images/payment/product-img-empty.svg',
      bankName: null,
      expiryDate: '12/33',
      expiryMonth: 12,
      expiryYear: 2033,
      id: '590925219668099073',
      isDefault: true,
      isExpired: false,
      maskedNumber: '411111XXXXXX1111',
      maskedNumberByStar: '**** **** **** 1111',
      type: 'Visa',
      typeIcon: '/icons/types-credit-card/visa-no-border.svg',
    },
    isShowCardList: true,
    setIsShowCardList: jest.fn(),
    setCurrentSelectedCard: jest.fn(),
    setOpenCreditCardFormInSavedList: jest.fn(),
  };

  useCreditCardContext.mockReturnValue(mockCreditCardContext);
  useReduxContext.mockReturnValue({
    reduxState: {
      cart: {
        cartTotals: {
          base_grand_total: 1000,
        },
      },
    },
  });

  const spyScrollTo = jest.fn();
  beforeEach(() => {
    Object.defineProperty(global.window, 'scrollTo', { value: spyScrollTo });
    spyScrollTo.mockClear();
  });

  test('Render card list | it should match with snapshot', async () => {
    const { asFragment } = render(<SaveCreditCardList {...props} />);
    await waitFor(() => {
      expect(asFragment(<SaveCreditCardList {...props} />)).toMatchSnapshot();
    });
  });
  test('cart total more than 10000 | it should show Cvv form', async () => {
    const props = {
      ...props,
      isShowCvvForm: true,
      isCardSelected: true,
    };
    useReduxContext.mockReturnValue({
      reduxState: {
        cart: {
          cartTotals: {
            base_grand_total: 11000,
          },
        },
      },
    });
    const { getByTestId, asFragment } = render(
      <SaveCreditCardList {...props} />,
    );
    await waitFor(() => {
      expect(getByTestId(TEST_ID.CVV_FORM)).toBeInTheDocument();
      expect(asFragment(<SaveCreditCardList {...props} />)).toMatchSnapshot();
    });
  });
  test(`when user click on input and leave | it should be show error message ${TEST_DATA.ERRORS_CVV.CARDCVV_REQUIRED}`, async () => {
    const props = {
      ...props,
      isShowCvvForm: true,
      isCardSelected: true,
    };
    useReduxContext.mockReturnValue({
      reduxState: {
        cart: {
          cartTotals: {
            base_grand_total: 11000,
          },
        },
      },
    });
    const { getByTestId } = render(<SaveCreditCardList {...props} />);
    const input = getByTestId(TEST_ID.INPUT_CVV_FORM);
    expect(input).toBeInTheDocument();
    fireEvent.click(input);
    fireEvent.blur(input);
    await waitFor(() => {
      expect(getByTestId(TEST_ID.CARD_CVV_ERROR)).toHaveTextContent(
        'payment.credit_card.form.errors.card_CVV_required',
      );
    });
  });
  test(`when user click on input invalid cvv | it should be show error message ${TEST_DATA.ERRORS_CVV.CARDCVV_INVALID}`, async () => {
    const props = {
      ...props,
      isShowCvvForm: true,
      isCardSelected: true,
    };
    useReduxContext.mockReturnValue({
      reduxState: {
        cart: {
          cartTotals: {
            base_grand_total: 11000,
          },
        },
      },
    });
    const { getByTestId } = render(<SaveCreditCardList {...props} />);
    const input = getByTestId(TEST_ID.INPUT_CVV_FORM);
    expect(input).toBeInTheDocument();
    fireEvent.input(input, { target: { value: '12' } });
    fireEvent.blur(input);
    await waitFor(() => {
      expect(getByTestId(TEST_ID.CARD_CVV_ERROR)).toHaveTextContent(
        'payment.credit_card.form.errors.card_CVV_invalid',
      );
    });
  });
  test(`when user input and leave | it should be show error message ${TEST_DATA.ERRORS_CVV.CARDCVV_REQUIRED}`, async () => {
    const props = {
      ...props,
      isShowCvvForm: true,
      isCardSelected: true,
    };
    useReduxContext.mockReturnValue({
      reduxState: {
        cart: {
          cartTotals: {
            base_grand_total: 11000,
          },
        },
      },
    });
    const { getByTestId } = render(<SaveCreditCardList {...props} />);
    const input = getByTestId(TEST_ID.INPUT_CVV_FORM);
    expect(input).toBeInTheDocument();
    fireEvent.click(input);
    fireEvent.blur(input);
    await waitFor(() => {
      expect(getByTestId(TEST_ID.CARD_CVV_ERROR)).toHaveTextContent(
        'payment.credit_card.form.errors.card_CVV_require',
      );
    });
  });
  test(`when user click add new credit card button | it should be call function handleOpenAddCardForm() `, async () => {
    const { getByTestId } = render(<SaveCreditCardList {...props} />);
    const addNewCard = getByTestId(TEST_ID.ADD_NEW_CARD);
    fireEvent.click(addNewCard);
    await waitFor(() => {
      expect(addNewCard).toBeInTheDocument();
    });
  });
  test(`when user click credit card item | it should be call function handleSelectCard() `, async () => {
    const { getAllByTestId } = render(<SaveCreditCardList {...props} />);
    const selectCard = getAllByTestId(
      'inf-viewSavedCreditCardListItem-Container',
    );
    fireEvent.click(selectCard[0]);
    await waitFor(() => {
      expect(selectCard[0]).toBeInTheDocument();
    });
  });
});
