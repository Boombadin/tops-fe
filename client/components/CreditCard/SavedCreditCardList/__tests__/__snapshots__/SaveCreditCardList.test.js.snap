// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CreditCard/SavedCreditCardList Render card list | it should match with snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="SavedCreditCardList__CardWithCvvFormWrap-sc-13xmdt6-0 kQxAJa"
    >
      <div
        class="SavedCreditCardList__CardWithCvvFormInside-sc-13xmdt6-1 jXJzZk"
      >
        <mock-creditcarditem
          card="[object Object]"
          data-testid="inf-viewSavedCreditCardListItem-Container"
          iscanbedelete="false"
          iscanbeselect="true"
          selectedcard="[object Object]"
        />
      </div>
      <div
        class="SavedCreditCardList__CardWithCvvFormInside-sc-13xmdt6-1 jXJzZk"
      >
        <mock-creditcarditem
          bankname="KCC"
          card="[object Object]"
          data-testid="inf-viewSavedCreditCardListItem-Container"
          iscanbedelete="false"
          iscanbeselect="true"
          selectedcard="[object Object]"
        />
      </div>
      <div
        class="SavedCreditCardList__AddNewCardWrap-sc-13xmdt6-6 bZZsWh"
        data-testid="btn-addSavedCreditCardList-AddNewCard"
      >
        <div
          class="SavedCreditCardList__AddNewCardInside-sc-13xmdt6-7 eRkAj"
        >
          <div
            class="SavedCreditCardList__AddNewCardIcon-sc-13xmdt6-8 fynLKy"
          >
            +
          </div>
          <p
            class="Text-sc-9p67zt-0 fIdxrL"
            color="#171722"
            display=""
            size="13"
            spacing="normal"
          >
            payment.credit_card.add-new-card
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`components/CreditCard/SavedCreditCardList cart total more than 10000 | it should show Cvv form 1`] = `
<DocumentFragment>
  <div>
    <div
      class="SavedCreditCardList__CardWithCvvFormWrap-sc-13xmdt6-0 kQxAJa"
    >
      <div
        class="SavedCreditCardList__CardWithCvvFormInside-sc-13xmdt6-1 jXJzZk"
      >
        <mock-creditcarditem
          card="[object Object]"
          data-testid="inf-viewSavedCreditCardListItem-Container"
          iscanbedelete="false"
          iscanbeselect="true"
          selectedcard="[object Object]"
        />
        <div
          data-testid="frm-viewSavedCreditCardList-CvvFormContainer"
        >
          <form
            data-testid="frm-viewSavedCreditCardList-CvvForm"
            id="2c2p-payment-form-cvv"
          >
            <div
              class="SavedCreditCardList__CardCvvWrap-sc-13xmdt6-2 dVOBKN"
            >
              <div
                class="SavedCreditCardList__CardCvvInside-sc-13xmdt6-3 bOFszE"
              >
                <div
                  class="SavedCreditCardList__CardInputCVV-sc-13xmdt6-4 hcdOjb"
                >
                  <input
                    autocomplete="off"
                    class="SavedCreditCardList__CardInput-sc-13xmdt6-5 gMhUuk"
                    data-encrypt="cvv"
                    data-testid="inp-formSavedCreditCardList-CardCVV"
                    id="cardCVV"
                    maxlength="3"
                    name="cardCVV"
                    placeholder="payment.credit_card.form.card_cvv_placeholder"
                    type="tel"
                    value=""
                  />
                </div>
                <mock-tooltipcvv />
              </div>
            </div>
          </form>
        </div>
      </div>
      <div
        class="SavedCreditCardList__CardWithCvvFormInside-sc-13xmdt6-1 jXJzZk"
      >
        <mock-creditcarditem
          bankname="KCC"
          card="[object Object]"
          data-testid="inf-viewSavedCreditCardListItem-Container"
          iscanbedelete="false"
          iscanbeselect="true"
          selectedcard="[object Object]"
        />
      </div>
      <div
        class="SavedCreditCardList__AddNewCardWrap-sc-13xmdt6-6 bZZsWh"
        data-testid="btn-addSavedCreditCardList-AddNewCard"
      >
        <div
          class="SavedCreditCardList__AddNewCardInside-sc-13xmdt6-7 eRkAj"
        >
          <div
            class="SavedCreditCardList__AddNewCardIcon-sc-13xmdt6-8 fynLKy"
          >
            +
          </div>
          <p
            class="Text-sc-9p67zt-0 fIdxrL"
            color="#171722"
            display=""
            size="13"
            spacing="normal"
          >
            payment.credit_card.add-new-card
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
